content/blog/2025-04-09-introducing-bloom-filters.md

```diff
@@ -0,0 +1,154 @@
+This bit-based approach, rather than full item allocation, makes bloom filters very space efficient with the trade off being potential false positives.
+
+Valkey-Bloom (BSD-3 licensed) is an official Valkey module that introduces bloom filters as a new data type to Valkey, providing both scalable and non-scalable variants.
+It is API compatible with the bloom filter command syntax of the official Valkey client libraries including valkey-py, valkey-java, valkey-go (as well as the equivalent Redis libraries).
+The module was developed by the following authors: Karthik Subbarao ([KarthikSubbarao](https://github.com/KarthikSubbarao)), Cameron Zack ([zackcam](https://github.com/zackcam)), Vanessa Tang ([YueTang-Vanessa](https://github.com/YueTang-Vanessa)), Nihal Mehta ([nnmehta](https://github.com/nnmehta)), and wuranxx ([wuranxx](https://github.com/wuranxx)).
+
+## Data type overview
+
+The "Bloom Object" is the main bloom data type structure. This is what gets created with any bloom filter creation command and this structure can act either as a "scaling bloom filter" or "non scaling bloom filter" depending on the user configuration.
+It consists of a vector of "Sub Filters" with length >= 1 in case of scaling and only 1 in case of non scaling.
+
+The "Sub Filter" is an inner structure which is created and used within the "Bloom Object". It tracks the capacity, number of items added, and an instance of a Bloom Filter (of the specified properties) from an external crate named bloomfilter.
+
+![bloom filter data type](/assets/media/pictures/bloomfilter_datatype.png)
+
+**Non Scaling**
+
+When non-scaling filters reach their capacity, if a user tries to add a new/unique item, an error is returned.
+You can create a non scaling bloom filter using `BF.RESERVE` or `BF.INSERT` commands.
+Example:
+```
+BF.RESERVE <filter-name> <error-rate> <capacity> NONSCALING
+``` 
+
+**Scaling**
+
+When scaling filters reach their capacity, if a user adds an item to the bloom filter, a new sub filter is created and added to the vector of sub filters.
+This new bloom sub filter will have a larger capacity (previous_bloomfilter_capacity * expansion_rate of the bloom filter).
+When checking whether an item exists on a scaled out bloom filter (`BF.EXISTS`/`BF.MEXISTS`), we look through each filter (from oldest to newest) in the sub filter vector and perform a check operation on each one.
+Similarly, to add a new item to the bloom filter, we check through all the filters to see if the item already exists and the item is added to the current filter if it does not exist.
+Any default creation as a result of `BF.ADD`, `BF.MADD`, `BF.INSERT` will be a scalable bloom filter.
+
+**Common Bloom filter properties**
+
+Capacity - The number of unique items that can be added before a bloom filter scales out occurs (in case of scalable bloom filters) or before any command which inserts a new item is rejected (in case of non scalable bloom filters).
```

Comment:
```suggestion
Capacity - The number of unique items that can be added before a bloom filter scales out occurs (in case of scalable bloom filters) or before any command which inserts a new item will return an error (in case of non scalable bloom filters).
```